// depcheck/dist/utils/linters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = parseLinter;

var _path = _interopRequireDefault(require("path"));

var _jsYaml = _interopRequireDefault(require("js-yaml"));

var _lodash = _interopRequireDefault(require("lodash"));

var _requirePackageName = _interopRequireDefault(require("require-package-name"));

var _ = require(".");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function parse(content) {
  try {
    return JSON.parse(content);
  } catch (error) {// not JSON format
  }

  try {
    return _jsYaml.default.safeLoad(content);
  } catch (error) {// not YAML format
  }

  try {
    return (0, _.evaluate)(`module.exports = ${content}`);
  } catch (error) {} // not valid JavaScript code
  // parse fail, return nothing


  return {};
}

function isLinterConfigAnAbsolutePath(specifier) {
  return _path.default.isAbsolute(specifier);
}

function isLinterConfigARelativePath(specifier) {
  return _lodash.default.startsWith(specifier, './') || _lodash.default.startsWith(specifier, '../');
}

function isLinterConfigFromAPlugin(specifier) {
  return _lodash.default.startsWith(specifier, 'plugin:');
}

function isLinterConfigFromAScopedModule(specifier) {
  return _lodash.default.startsWith(specifier, '@');
}

function isLinterConfigFromAFullyQualifiedModuleName(specifier, prefix) {
  return _lodash.default.startsWith(specifier, prefix);
}

function resolvePresetPackage(flavour, preset, rootDir) {
  // inspired from https://github.com/eslint/eslint/blob/5b4a94e26d0ef247fe222dacab5749805d9780dd/lib/config/config-file.js#L347
  if (isLinterConfigAnAbsolutePath(preset)) {
    return preset;
  }

  if (isLinterConfigARelativePath(preset)) {
    return _path.default.resolve(rootDir, preset);
  }

  const _ref = isLinterConfigFromAPlugin(preset) ? {
    prefix: `${flavour}-plugin-`,
    specifier: preset.substring(preset.indexOf(':') + 1)
  } : {
    prefix: `${flavour}-config-`,
    specifier: preset
  },
        prefix = _ref.prefix,
        specifier = _ref.specifier;

  if (isLinterConfigFromAScopedModule(specifier)) {
    const scope = specifier.substring(0, specifier.indexOf('/'));
    const module = specifier.substring(specifier.indexOf('/') + 1);

    if (isLinterConfigFromAFullyQualifiedModuleName(module, prefix)) {
      return specifier;
    }

    return `${scope}/${prefix}${module}`;
  }

  if (isLinterConfigFromAFullyQualifiedModuleName(specifier, prefix)) {
    return specifier;
  }

  return `${prefix}${specifier}`;
}

function loadConfig(preset, rootDir) {
  const presetPath = _path.default.isAbsolute(preset) ? preset : _path.default.resolve(rootDir, 'node_modules', preset);

  try {
    return require(presetPath); // eslint-disable-line global-require
  } catch (error) {
    return {}; // silently return nothing
  }
}
/**
 * Brings package name to correct format based on prefix
 * @param {string} name The name of the package.
 * @param {string} prefix Can be either "eslint-plugin", "eslint-config" or "eslint-formatter"
 * @returns {string} Normalized name of the package
 * @private
 * @see {@link https://github.com/eslint/eslint/blob/faf3c4eda0d27323630d0bc103a99dd0ecffe842/lib/util/naming.js#L25 ESLint}
 */


function normalizePackageName(name, prefix) {
  let normalizedName = name;

  const convertPathToPosix = p => _path.default.normalize(p).replace(/\\/g, '/');
  /**
   * On Windows, name can come in with Windows slashes instead of Unix slashes.
   * Normalize to Unix first to avoid errors later on.
   * https://github.com/eslint/eslint/issues/5644
   */


  if (normalizedName.indexOf('\\') > -1) {
    normalizedName = convertPathToPosix(normalizedName);
  }

  if (normalizedName.charAt(0) === '@') {
    /**
     * it's a scoped package
     * package name is the prefix, or just a username
     */
    const scopedPackageShortcutRegex = new RegExp(`^(@[^/]+)(?:/(?:${prefix})?)?$`);
    const scopedPackageNameRegex = new RegExp(`^${prefix}(-|$)`);

    if (scopedPackageShortcutRegex.test(normalizedName)) {
      normalizedName = normalizedName.replace(scopedPackageShortcutRegex, `$1/${prefix}`);
    } else if (!scopedPackageNameRegex.test(normalizedName.split('/')[1])) {
      /**
       * for scoped packages, insert the prefix after the first / unless
       * the path is already @scope/eslint or @scope/eslint-xxx-yyy
       */
      normalizedName = normalizedName.replace(/^@([^/]+)\/(.*)$/, `@$1/${prefix}-$2`);
    }
  } else if (normalizedName.indexOf(`${prefix}-`) !== 0) {
    normalizedName = `${prefix}-${normalizedName}`;
  }

  return normalizedName;
}

function checkConfig(flavour, config, rootDir) {
  const parser = (0, _.wrapToArray)(config.parser);
  const plugins = (0, _.wrapToArray)(config.plugins).map(plugin => normalizePackageName(plugin, `${flavour}-plugin`));
  const presets = (0, _.wrapToArray)(config.extends).filter(preset => preset !== `${flavour}:recommended`).map(preset => resolvePresetPackage(flavour, preset, rootDir));
  const presetPackages = presets.filter(preset => !_path.default.isAbsolute(preset)).map(_requirePackageName.default);
  const presetDeps = (0, _lodash.default)(presets).map(preset => loadConfig(preset, rootDir)).map(presetConfig => checkConfig(flavour, presetConfig, rootDir)).flatten().value();
  return _lodash.default.union(parser, plugins, presetPackages, presetDeps);
}

function parseLinter(flavour, content, filename, deps, rootDir) {
  const basename = _path.default.basename(filename);

  const filenameRegex = new RegExp(`^\\.?${flavour}(rc)?(\\.json|\\.js|\\.yml|\\.yaml)?$`);

  if (filenameRegex.test(basename)) {
    const config = parse(content);
    return checkConfig(flavour, config, rootDir);
  }

  return [];
}

module.exports = exports.default;
